File: evaluation/S0/c/Medium/1343-count-good-meals.c
// Source: https://leetcode.com/problems/count-good-meals/   |   Difficulty: Medium
//
// Problem Description:
// A good meal is a meal that contains exactly two different food items with a sum of deliciousness equal to a power of two.
//
// You can pick any two different foods to make a good meal.
//
// Given an array of integers deliciousness where deliciousness[i] is the deliciousness of the i​​​​​​th​​​​​​​​ item of food, return the number of different good meals you can make from this list modulo 109 + 7.
//
// Note that items with different indices are considered different even if they have the same deliciousness value.
//
// Example:
// Input: deliciousness = [1,3,5,7,9]
// Output: 4
// Explanation: The good meals are (1,3), (1,7), (3,5) and, (7,9).
// Their respective sums are 4, 8, 8, and 16, all of which are powers of 2.
//
// Constraints:
// 1 <= deliciousness.length <= 105
// 	0 <= deliciousness[i] <= 220
//

int countPairs(int* deliciousness, int deliciousnessSize) {
    #define MOD 1000000007
    #define MAX_POWER 22

    int count[1 << (MAX_POWER + 1)];
    for (int i = 0; i < (1 << (MAX_POWER + 1)); i++) {
        count[i] = 0;
    }

    long long result = 0;
    for (int i = 0; i < deliciousnessSize; i++) {
        for (int power = 0; power <= MAX_POWER; power++) {
            int target = (1 << power) - deliciousness[i];
            if (target >= 0 && target < (1 << (MAX_POWER + 1))) {
                result = (result + count[target]) % MOD;
            }
        }
        count[deliciousness[i]]++;
    }

    return (int)result;     
}

File: evaluation/S1/python/Medium/1343-count-good-meals.py
# Source: https://leetcode.com/problems/count-good-meals/   |   Difficulty: Medium
#
# Problem Description:
# A good meal is a meal that contains exactly two different food items with a sum of deliciousness equal to a power of two.
#
# You can pick any two different foods to make a good meal.
#
# Given an array of integers deliciousness where deliciousness[i] is the deliciousness of the i​​​​​​th​​​​​​​​ item of food, return the number of different good meals you can make from this list modulo 109 + 7.
#
# Note that items with different indices are considered different even if they have the same deliciousness value.
#
# Example:
# Input: deliciousness = [1,3,5,7,9]
# Output: 4
# Explanation: The good meals are (1,3), (1,7), (3,5) and, (7,9).
# Their respective sums are 4, 8, 8, and 16, all of which are powers of 2.
#
# Constraints:
# 1 <= deliciousness.length <= 105
# 	0 <= deliciousness[i] <= 220
#
# Helpful references (internal KB):
# - Linear Diophantine Equation (number, interval, enumeration)
#   • When to use: Use when you need to find or count all integer solutions (x, y) for a linear Diophantine equation ax + by = c, especially when solutions are constrained within specific intervals for x and y.
#   • Idea: This method finds all integer solutions to a linear Diophantine equation ax + by = c within given x and y intervals. It first finds a particular solution, then uses properties of the general solution to determine the valid range for x, and enumerates solutions in O(N) time where N is the number of solutions.
#   • Invariants: The current (x, y) pair always satisfies the linear Diophantine equation ax + by = c.; The current x value is an integer and satisfies the effective bounds l_x <= x <= r_x.
#   • Tips: Pre-divide coefficients a, b, c by gcd(a, b) to simplify the equation.; Use the Extended Euclidean Algorithm to find an initial particular solution (x0, y0).
#   • Pitfalls: Failing to check if c is divisible by gcd(a, b) before proceeding, leading to incorrect results.; Incorrectly applying the shifting logic for x and y bounds, resulting in an invalid solution range.
# - Disjoint Sparse Table tutorial (array, ds, number, sparse-table, counting)
#   • When to use: Use when you need to answer range queries on a static array in O(1) time, after an initial O(N log N) preprocessing phase. It is particularly efficient for associative operations like sum, min, or max.
#   • Idea: A Disjoint Sparse Table is a data structure that precomputes results for all possible range queries on a static array using a hierarchical block decomposition. It achieves O(N log N) construction time and space complexity, enabling O(1) query time for associative operations.
#   • Invariants: ST[k][i] stores the result for a range of length 2^k starting at i within its block.; For any level k, blocks are disjoint, but ranges within ST[k] can overlap with ranges in ST[k-1].
#   • Tips: Precompute log2 values or highest set bit for efficient level determination.; Ensure the operation (e.g., sum, min, max) is associative for correct combination.
#   • Pitfalls: Not suitable for arrays that require updates, as it's a static data structure.; Only works for associative operations; non-associative operations cannot be combined this way.
# - String Hashing (string, array, hashing, enumeration)
#   • When to use: Use string hashing to efficiently compare substrings or patterns within a larger text, or to quickly check for duplicate strings. It's ideal for problems requiring fast equality checks on string segments.
#   • Idea: String hashing converts a string or substring into a numerical hash value, allowing for O(1) comparison of substrings after O(N) precomputation. It typically uses polynomial hashing with a prime base and modulus to minimize collisions.
#   • Invariants: The hash value for the prefix S[0...i] is correctly computed and stored, based on the chosen polynomial hashing scheme.; The base 'p' and modulus 'm' remain constant throughout the hash computation for consistent results.
#   • Tips: Choose a large prime number for the base 'p' (e.g., 31, 53) to reduce collisions.; Select a large prime number for the modulus 'm' (e.g., 10^9+7, 10^9+9) to further minimize collision probability.
#   • Pitfalls: Avoid using 2^64 as a modulus, as specific collision-generating strings exist for this choice.; A small modulus 'm' or base 'p' drastically increases the likelihood of hash collisions.
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

class Solution(object):
    def countPairs(self, deliciousness):
        """
        :type deliciousness: List[int]
        :rtype: int
        """
        MOD = 10**9 + 7
        count = {}
        result = 0
        
        max_val = max(deliciousness)
        max_power = 1
        while max_power < 2 * max_val:
            max_power *= 2
        
        for val in deliciousness:
            power = 1
            while power <= max_power:
                complement = power - val
                if complement in count:
                    result = (result + count[complement]) % MOD
                power *= 2
            
            if val in count:
                count[val] += 1
            else:
                count[val] = 1
        
        return result